### DISCUSSION TITLE
[Feature Request] Global Environmental Variables

### DISCUSSION THREAD

**Discussion by MrTheoW** [louislam/dockge#133] [Feature Request]
It would be awesome to be able to set some global env. variables.
Like things that doesn't change as:

- Time Zone
- PGID & PUID
- Docker Data Dir Location

It would be nice to be able to name a variable and give it a value so the variable could be used in the stacks / containers.

**Comment by TheRealGramdalf** [comment]
I recently came across [fragments](https://docs.docker.com/compose/compose-file/10-fragments/) and [extensions](https://docs.docker.com/compose/compose-file/11-extension/) - I think an integration of this would be the best way to implement this.

Here's my thoughts on how this might end up working:
- Global configurations are kept in the primary `dockge` `compose.yaml`
- Other stacks can reference these global config options, and optionally override them on a per-stack basis
- App templates may optionally be integrated into this setup - for example with TZ, containers can reference the `dockge` compose file in order to fill that environment variable
- If a container uses a different name for a variable, it can be set with something like `- CUSTOM_TZ=$TZ` (using basic shell syntax, I'm not exactly sure of the proper compose syntax for variables)
- An added bonus would be a simple UI to manage allocations - say you have two stacks, stack A and stack B. Stack A supports time zone and PG/UID, but stack B only supports time zone - a drag n drop/checklist UI to select which variables a container pulls from `dockge` would be nice

Since `dockge` in the vast majority of cases defines your docker environment, it makes sense that it would contain these configuration options.

**Reply by supacoda10** [reply]
Or maybe something like Dokemon (https://dokemon.dev) does would be nice too. You can define environments in it and variables under an environment. Then assign the environment to the node. All variables in that environment are then set for all stacks in that node. This is useful in our company where we have QA, UAT, Production environments.

I like Dockge because it shows everything related to a stack on one page which is cool. Would be nice to have something like this in Dockge.

**Comment by MikyWoW** [comment]
You can also use simple `.env` file in same directory as `compose.yaml` for per stack enviroment variables.

With global env then run compose command like this `docker compose --env-file ./../global.env --env-file .env up -d`

https://docs.docker.com/compose/environment-variables/set-environment-variables/

**Comment by FallenLeave91** [comment]
An option to change the data dir would be apreciated since i store my stack under ``/app`` and don´t want to move it. This is in conflict with the default config dir in ``/app/data``

**Reply by louislam** [reply]
Good point. I shouldn't put the whole dockge application in `/app`, as the name is a bit common.

**Comment by trythatagain** [comment]
I use OMV and the compose plugin, which is very nice, but I like new things as well, so I'm enjoying Dockge. But I am missing the global env file I have with the compose plugin, so as mrTheoW mentions, I have my most used directories in there, TZ, PUID etc, things that don't really change.
It's a really nice feature, plus as mentioned here also you can use a local env as well.

I'd quite like to create a global env file, stored either in the root of the stacks directory, or wherever really, and then an option in the settings to state the path, and perhaps a checkbox to state a global env file is available, or if it's easier have a 0 byte file auto created and called every time Dockge does and up.

**Reply by synapseattack** [reply]
Also a fellow OMV user and I too miss having a globals file for variables that can be used.

**Comment by richard-edwards** [comment]
I've been using the .env file in the same folder as compose.yml and it's working well. Would be nice to be able to quickly edit that in the interface along with the compose file.

**Reply by trythatagain** [reply]
I get that, but it means you've got to copy .env every time you create a new stack, whereas if you had a global .env and Dockge knows the path, when it does the 'compose up'  it includes that as default. One less thing to have to remember to do.

I just checked using OMV compose plugin, which has a default .env for each stack dir, it will startup if the ,env file is 0 bytes. So I assume if a global .env was created when Dockge first runs, with a description in the file, then when the 'compose up' includes the global .env in it's run argument then everything will run as wanted and those of use who want a global .env file will be extremely happy :)

**Comment by Anaxagoras-bc** [comment]
putting in a vote for this.

**Comment by syko9000** [comment]
I was rearranging my stacks and now I really want this... I'm going to implement this @louislam and you can decide if you accept it. Might be a few days or the weekend though.

**Reply by louislam** [reply]
Sure. I have the design in my mind, see if you agree or any improvement:

- The global env file stored at: `/opt/stacks/global.env`
- Add a new page "Global .env" in the "Settings" page for editing global.env
- Add a checkbox in the stack page (edit mode) to enable/disable global .env
- When enable/disable, modify the yaml:

```
services:
  webapp:
    image: my-webapp-image
    env_file:          
      - ../global.env      # Enable = add this line   Disable = remove this line
```

**Reply by syko9000** [reply]
I think that's fine, however, `env_file` only passes those vars into the container, they don't appear to be available to be used in the compose.yaml file like the `.env` is. For them to be usable in compose the files need to be added to the `docker compose up -d` like @MikyWoW mentioned above. Then the checkbox in the stack page to pass those global vars (and local vars) verbatim to the container is a nice extra.

**Reply by trythatagain** [reply]
Surely if you have a checkbox in the stack-->edit page, there is no need to a env_file value?
Could you  also create a 0 byte global,env file when spinning up dockge, and then just either have a checkbox in the stack edit page, or even default to just using the global.env when doing the -up.

I use OMV 6, and the compose plugin uses a global.env, and I only have a checkbox for when I'm creating the global.env page, I never have to worry about checkbox in the stack edit window, or having to include global.env variable in my stack.
I really wish I could still program, but I'm to old to learn again :)

**Reply by syko9000** [reply]
I have the UI in place and it works with the docker compose up call but the PORT detection through the env parsing library being used needs augmented to read the two files and to process each file's values through the variables known in the files. It's actually complicated...

global
```
PORT=1
PORT=${PORT}2
PORT=${PORT}3
```
local
```
PORT=${PORT}4
PORT=${PORT}5
PORT2=${PORT}6
PORT2=${PORT2}7
```
then a service with an environment var of `PORT=$PORT` and `PORT2=$PORT2` for some reason yields `PORT=1235` and `PORT2=12367`

In the `ports:` section it comes back as 
<img width="269" alt="Screenshot 2024-01-06 at 10 12 48 PM" src="https://github.com/louislam/dockge/assets/193088/c722f3bd-2f2d-4613-aa28-d91d7528124c">

So it seems iterative inside a file, but first pass when applying to the next file followed iteratively. Very weird.

Edge cases are a nightmare. Hope to get it tomorrow.

**Reply by louislam** [reply]
I think they are really edge cases and maybe even invalid? Because key should be unique in my opinion.

And I also don't expected that the envsubst operation is happening inside .env files. (Need to verify how docker-compose does)

We can assume that the key must be unique and the local one should have the highest priority.

Otherwise, it will be super complicated.

**Reply by trythatagain** [reply]
I think you're correct, the local.env file should take priority. I don't program, but does compose up, not do this for you? Especially if you're adding both env file names to the compose up command?

**Reply by syko9000** [reply]
Docker compose definitely allows for the substitutions within the .env file and it didn't complain with my test case at all having duplicate keys. Just like in BASH you can redefine vars all you want.

<img width="686" alt="Screenshot 2024-01-07 at 11 07 51 AM" src="https://github.com/louislam/dockge/assets/193088/b8a3f9e7-7ee7-4b46-8b2e-38b9f857bf92">

Even further it allows for nested things like `AUX_PORT=${PORT:-${DEFAULT_PORT:-8080}}` within the same file (ie putting this in global.env and defining DEFAULT_PORT in .env does not cause AUX_PORT to re-evaluate, luckily)

**Reply by syko9000** [reply]
@trythatagain, yes, docker compose handles all of it as far as the running containers are concerned. But we have to emulate what compose does in certain situations so that the UI matches the real result.

Unless, don't show the port at all if the container isn't running and if the container is running run a `docker container inspect ...` and pull the port number that was resolved by compose out for the UI?

**Comment by aviellg** [comment]
for the interface maybe take inspiration from yacht project , it will be better if you can create new veriables

**Comment by bverkron** [comment]
I would also love this feature. I have many containers that share certain env variables and it's tedious to always copy / paste them into individual stack configs and repeat the process if values change.

**Comment by hilsonp** [comment]
Thank you @louislam for the nice project. This is immensely useful !
And thank you @syko9000 for the contribution.
Impatient to get the 1.5 release. 
In the mean time, is there any way to beta test it ?
Do you have a little HowTo build the dockge docker image ?

**Reply by RafaelSchridi** [reply]
Would also love to test this, is there a docker tag for the pull request?

**Reply by RafaelSchridi** [reply]
For anyone interested
```
git clone https://github.com/syko9000/dockge.git
git checkout master.global.env
npm install (you need node ~18.17.1, you can use nvm for this)
npm run build:frontend 
sudo docker buildx build -t louislam/dockge:pr387 --target release -f ./docker/Dockerfile .
```
Now you can use the image `louislam/dockge:pr387` to rebuild your container

**Comment by raphhgg** [comment]
Any progress on this? Much needed feature!

**Reply by zdhoward** [reply]
No suggested solutions was working with my setup so I just ended up making a template stack folder with a symbolic link to a .env file and I duplicate it and rename it to make new stacks now. Pretty simple to setup, just wish I could do it all from inside dockge.

My stacks dir:
```
/opt/stacks/
- .env (that contains all the global vars I want)
- new.sh (this is the script I use to make new stacks)
- template/
-- compose.yaml (empty but needs to be there to show up in dockge)
-- .env (symbolic link to the .env file in the parent directory
```

the command to make the symbolically linked .env file:
`ln -s /opt/stacks/.env /opt/stacks/template/.env`

and my new.sh script:
`cp -r /opt/stacks/template/ /opt/stacks/$1/`

then when I want to make a new stack (named newStackName) I run a command like:
`/opt/stacks/new.sh newStackName`

after that you can just hop into dockge and when you edit the .env file in the newStackName stack, it will reflect those changes in every stack that is symbolically linked. Just remember to update the stacks for new changes to the .env to take effect.

**Reply by benblustey** [reply]
Symbolic link is a good call. Worked perfectly for me.

**Reply by bverkron** [reply]
Can you have an additional .env file inside each app’s folder or are you restricted to a single global .env?

**Reply by zdhoward** [reply]
This way is just using the symbolic link, so it only refers back to the global .env
If you just put a normal .env in the apps folder then it won't refer to the global .env at all, but this is the same as just adding a new stack in dockge directly

**Reply by bverkron** [reply]
Right, but as wondering if docker had a mechanism to allow for multiple .env files, so you could have the global one and app specific one, but I don’t think so. Thought maybe you found way :)

**Reply by TheRealGramdalf** [reply]
There is in fact a way to do that - you can either edit the docker daemon settings to add additional names to search (like `.env.local`), or add `./.env.local` to an `environment_files` directive

**Comment by ak4zh** [comment]
I ending up doing this, so /opt/stacks/.env is now available globally to all containers. 
```yaml
services:
  dockge:
    image: louislam/dockge:1
    restart: unless-stopped
    ports:
      - 5002:5001
    env_file:
      - .env
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./data:/app/data
      - /opt/stacks:/opt/stacks
```

```env
DOCKGE_STACKS_DIR=/opt/stacks
TZ=Asia/Kolkata
MEDIA_DIRECTORY=/mnt/storage/media
```
`sudo ln -s /opt/stacks/.env /opt/dockge/.env`

When I add new variables, I do this:
```bash
cd /opt/dockage
docker compose down
docker compose up -d
```
